kernel/libs/aster-bigtcp/src/iface/poll.rs

```diff
@@ -0,0 +1,473 @@
+        match dst_addr {
+            IpAddress::Ipv4(dst_addr) => self
+                .iface_cx
+                .ipv4_addr()
+                .is_some_and(|addr| addr == dst_addr),
+        }
+    }
+}
+
+impl<'a, E> PollContext<'a, E> {
+    pub(super) fn poll_egress<D, Q>(&mut self, device: &mut D, mut dispatch_phy: Q)
+    where
+        D: Device + ?Sized,
+        Q: FnMut(&Packet, &mut Context, D::TxToken<'_>),
+    {
+        while let Some(tx_token) = device.transmit(self.iface_cx.now()) {
+            if !self.dispatch_ipv4(tx_token, &mut dispatch_phy) {
+                break;
+            }
+        }
+    }
+
+    fn dispatch_ipv4<T, Q>(&mut self, tx_token: T, dispatch_phy: &mut Q) -> bool
+    where
+        T: TxToken,
+        Q: FnMut(&Packet, &mut Context, T),
+    {
+        let (did_something_tcp, tx_token) = self.dispatch_tcp(tx_token, dispatch_phy);
+
+        let Some(tx_token) = tx_token else {
+            return did_something_tcp;
+        };
+
+        let (did_something_udp, _tx_token) = self.dispatch_udp(tx_token, dispatch_phy);
+
+        did_something_tcp || did_something_udp
+    }
+
+    fn dispatch_tcp<T, Q>(&mut self, tx_token: T, dispatch_phy: &mut Q) -> (bool, Option<T>)
+    where
+        T: TxToken,
+        Q: FnMut(&Packet, &mut Context, T),
+    {
+        let mut tx_token = Some(tx_token);
+        let mut did_something = false;
+
+        for socket in self.tcp_sockets.iter() {
+            if !socket.need_dispatch(self.iface_cx.now()) {
+                continue;
+            }
+
+            // We set `did_something` even if no packets are actually generated. This is because a
+            // timer can expire, but no packets are actually generated.
+            did_something = true;
+
+            let mut deferred = None;
+
+            let reply = socket.dispatch(self.iface_cx, |cx, ip_repr, tcp_repr| {
+                let mut this = PollContext::new(cx, self.tcp_sockets, self.udp_sockets);
+
+                if !this.is_unicast_local(ip_repr.dst_addr()) {
+                    dispatch_phy(
+                        &Packet::new(ip_repr.clone(), IpPayload::Tcp(*tcp_repr)),
+                        this.iface_cx,
+                        tx_token.take().unwrap(),
+                    );
+                    return None;
+                }
+
+                if !socket.can_process(tcp_repr.dst_port) {
+                    return this.process_tcp(ip_repr, tcp_repr);
+                }
+
+                // We cannot call `process_tcp` now because it may cause deadlocks. We will copy
+                // the packet and call `process_tcp` after releasing the socket lock.
+                deferred = Some((ip_repr.clone(), {
+                    let mut data = vec![0; tcp_repr.buffer_len()];
+                    tcp_repr.emit(
+                        &mut TcpPacket::new_unchecked(data.as_mut_slice()),
+                        &ip_repr.src_addr(),
```

Comment:
`tcp_repr.emit` _writes_ a TCP packet into `data`, so we are not building a TCP packet from a 0-initialized array, we are creating a new TCP packet with the contents in `tcp_repr`.

This is theoretically possible because TCP connections can be self-connecting, see [here](https://totozhang.github.io/2016-01-11-tcp-self-connection/) for an example. Although I haven't experimented whether it is possible to create such self-connecting TCP connections in Asterinas, I think it is just better to keep the logic here.

Reply:
Thanks for your explanation!